File: app/form_validator.py
"""Validador de un formulario de registro sencillo."""

import re

EMAIL_PATTERN = re.compile(r"^[\w.+-]+@[\w-]+\.[a-zA-Z]{2,}$")

MIN_USERNAME_LENGTH = 3
MIN_PASSWORD_LENGTH = 8
MIN_AGE = 0
MAX_AGE = 120


def validate_form(username, email, password, age):
    """Devuelve la lista de errores encontrados en los datos del formulario."""
    errors = []

    if not username or len(username.strip()) < MIN_USERNAME_LENGTH:
        errors.append(f"El nombre de usuario debe tener al menos {MIN_USERNAME_LENGTH} caracteres.")

    if not email or not EMAIL_PATTERN.match(email):
        errors.append("El correo electrónico no tiene un formato válido.")

    if not password or len(password) < MIN_PASSWORD_LENGTH:
        errors.append(f"La contraseña debe tener al menos {MIN_PASSWORD_LENGTH} caracteres.")

    if age < MIN_AGE or age > MAX_AGE:
        errors.append(f"La edad debe estar entre {MIN_AGE} y {MAX_AGE} años.")

    return errors


def is_valid(username, email, password, age):
    """Indica si el formulario no contiene errores."""
    return len(validate_form(username, email, password, age)) == 0

File: src/main/java/com/cicdlabs/webform/FormValidator.java
package com.cicdlabs.webform;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.regex.Pattern;

/**
 * Valida los datos ingresados en el formulario web de registro.
 */
public class FormValidator {

    private static final Pattern EMAIL_PATTERN =
            Pattern.compile("^[\\w.+-]+@[\\w-]+\\.[a-zA-Z]{2,}$");

    private static final int MIN_USERNAME_LENGTH = 3;
    private static final int MIN_PASSWORD_LENGTH = 8;
    private static final int MIN_AGE = 0;
    private static final int MAX_AGE = 120;

    public List<String> validate(RegistrationForm form) {
        List<String> errors = new ArrayList<>();

        if (form.getUsername() == null || form.getUsername().trim().length() < MIN_USERNAME_LENGTH) {
            errors.add("El nombre de usuario debe tener al menos " + MIN_USERNAME_LENGTH + " caracteres.");
        }

        if (form.getEmail() == null || !EMAIL_PATTERN.matcher(form.getEmail()).matches()) {
            errors.add("El correo electrónico no tiene un formato válido.");
        }

        if (form.getPassword() == null || form.getPassword().length() < MIN_PASSWORD_LENGTH) {
            errors.add("La contraseña debe tener al menos " + MIN_PASSWORD_LENGTH + " caracteres.");
        }

        if (form.getAge() < MIN_AGE || form.getAge() > MAX_AGE) {
            errors.add("La edad debe estar entre " + MIN_AGE + " y " + MAX_AGE + " años.");
        }

        return Collections.unmodifiableList(errors);
    }

    public boolean isValid(RegistrationForm form) {
        return validate(form).isEmpty();
    }
}
